docs/guide/en/schema/typecasting.md

```diff
@@ -0,0 +1,293 @@
+
+Here is an example how to configure lazy type casting for the `array`, `json` and `structured` database types.
+
+```php
+use Yiisoft\Db\Constant\ColumnType;
+use Yiisoft\Db\Pgsql\Column\ColumnFactory;
+use Yiisoft\Db\Pgsql\Connection;
+use Yiisoft\Db\Schema\Column\ArrayLazyColumn;
+use Yiisoft\Db\Schema\Column\StructuredLazyColumn;
+use Yiisoft\Db\Schema\Data\JsonLazyArray;
+
+$columnFactory = new ColumnFactory(
+    'columnClassMap' => [
+        ColumnType::ARRAY => ArrayLazyColumn::class, // converts values to `LazyArray` objects 
+        ColumnType::JSON => JsonLazyColumn::class, // converts values to `JsonLazyArray` objects
+        ColumnType::STRUCTURED => StructuredLazyColumn::class, // converts values to `StructuredLazyArray` objects
+    ],
+);
+
+// Create a database connection with the custom column factory
+$db = new Connection($pdoDriver, $schemaCache, $columnFactory);
+
+/** @var JsonLazyArray $tags `tags` column is of database type `json` */
+$tags = $db->getTableSchema('customer')->getColumn('tags')->phpTypecast($row['tags']);
+
+foreach ($tags as $tag) {
+    echo $tag;
+}
+```
+
+## Structured data types
+
+Some databases support structured data types, such as `composite` types in Postgres. To cast a structured data type to
+a custom class, you need to create a column class which extends `AbstractStructuredColumn` and override 
+the `phpTypecast()` method.
+
+For example if `currency_money` is a defined composite type in Postgres as follows:
+
+```sql
+CREATE TYPE currency_money AS (
+    value DECIMAL(10,2),
+    currency_code CHAR(3)
+);
+```
+
+you can create `MyStructuredColumn` column class to cast the value to `CurrencyMoney` class.
+
+```php
+use Yiisoft\Db\Expression\Expression;
+use Yiisoft\Db\Expression\ExpressionInterface;
+use Yiisoft\Db\Pgsql\Data\StructuredParser;
+
+class MyStructuredColumn extends AbstractStructuredColumn
+{
+    /**
+    * @param string|null $value
+    */
+    public function phpTypecast(mixed $value): CurrencyMoney|StructuredLazyArray|null
+    {
+        if (is_string($value)) {
+            $value = new StructuredLazyArray($value, $this->getColumns());
+        
+            return match ($this->getDbType()) {
+                'currency_money' => new CurrencyMoney(...$value->getValue()),
+                default => $value,
+            };
+        }
+        
+        return $value;
+    }
+}
+
+class CurrencyMoney implements \JsonSerializable, \IteratorAggregate
+{
+    public function __construct(
+        private float $value,
+        private string $currencyCode = 'USD',
+    ) {
+    }
+
+    public function getCurrencyCode(): string
+    {
+        return $this->currencyCode;
+    }
+
+    public function getValue(): float
+    {
+        return $this->value;
+    }
+
+    /**
+     * `JsonSerializable` interface is implemented to convert the object to a database representation.
+     */
+    public function jsonSerialize(): array
```

Comment:
May be to add another example with `dbTypecast()` . This allow to keep DTO clear, without infrastructure things.

Reply:
This is how to convert object to array, then the array is typecasting and converting to string representation.

Also possible way to declare the properties as public.

https://github.com/yiisoft/db/blob/adf61039ab6dce56cfc08a2c1c3818df387297b5/src/Helper/DbArrayHelper.php#L235-L251